system/Helpers/array_helper.php

```diff
@@ -218,3 +218,49 @@ function array_flatten_with_dots(iterable $array, string $id = ''): array
         return $flattened;
     }
 }
+
+if (! function_exists('array_group_by')) {
+    /**
+     * Groups all rows by their index values. Result's depth equals number of indexes
+     *
+     * @param array $array        Data array (i.e. from query result)
+     * @param array $indexes      Indexes to group by. Dot syntax used. Returns $array if empty
+     * @param bool  $includeEmpty If true, null and '' are also added as valid keys to group
+     *
+     * @return array Result array where rows are grouped together by indexes values.
+     */
+    function array_group_by(array $array, array $indexes, bool $includeEmpty = false): array
+    {
+        if (empty($indexes)) {
+            return $array;
+        }
+
+        $result = [];
+
+        foreach ($array as $row) {
+            $currentLevel = &$result;
```

Comment:
Why would you need to use reference here? Is using the `$result` directly would not be okay? I'm asking so as to fix the phpstan error on L247.

Reply:
Yes, the reference is indead necessary. I want to work with the actual $result array, not just with a copy of it (difference between "assign by value" and "assign by reference"). Otherwise `$currentLevel` and `$result` would be 2 different arrays.

My unit tests fail if I declare `$currentLevel = $result`.

Reply:
It seems that the code also works with
```php
if (! isset($currentLevel[$value])) {
    $currentLevel[$value] = [];
}
```

Reply:
Though slower, can you try using `array_reduce()` instead of `foreach`?

Reply:
I think that makes the code more difficult to read. I would need to include `$result` and `$indexes` to the anonymous function.